File: 1048-longest-string-chain/1048-longest-string-chain.py
class Solution:
    def longestStrChain(self, words: List[str]) -> int:
        k=collections.defaultdict(set)
        out={i:1 for i in words}
        for i in words:
            k[len(i)].add(i)
        for i in sorted(k.keys(),reverse=True):
            if i-1 in k:
                for w in k[i]:
                    for j in range(len(w)):
                        s=w[:j]+w[j+1:]
                        if s in k[i-1]:
                            out[s]=max(out[s],out[w]+1)
        return max(out.values())
                    

File: 1048-longest-string-chain/1048-longest-string-chain.cpp
class Solution {
public:
    int longestStrChain(vector<string>& ss) {
        map<int,vector<string>,greater<int>>k;
        map<string,int>ans;
        int out=1;
        for(string s:ss)
        {
            ans[s]=1;
            k[s.size()].push_back(s);
        }
        for(auto it:k)
        {
            if(k.find(it.first-1)!=k.end())
            {
                for(string s:it.second)
                {
                    for(int j=0;j<s.size();j++)
                    {
                        string l=s.substr(0,j);
                        string r=s.substr(j+1);
                        string temp=l+r;
                        if(ans.find(temp)!=ans.end())
                        {
                            ans[temp]=max(ans[temp],ans[s]+1);
                            out=max(out,ans[temp]);
                        }
                    }
                }
            }
        }
        return out;
        
    }
};